clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3177,6 +3177,8 @@ static void handleWorkGroupSize(Sema &S, Decl *D, const ParsedAttr &AL) {
 
   ASTContext &Ctx = S.getASTContext();
 
+  assert(YDimExpr && "YDimExpr cannot be NULL");
+  assert(ZDimExpr && "ZDimExpr cannot be NULL");
```

Comment:
This is a bit different than the previous suggestion:
```
Agreed; I think the Klokwork report is a false positive, but we could eliminate it (hopefully) and improve the readability of the code by using an assert within the lambda to assert that the only way to call the lambda is for an attribute that can elide the Y and Z arguments. Then we can remove the return of nullptr from the lambda and simplify that code a bit.
```
I think the `assert` should be in the `SetDefault` lambda. It should be asserting:
```
(AL.getKind() == ParsedAttr::AT_ReqdWorkGroupSize && AL.hasScope() &&
         (AL.getScopeName()->isStr("sycl") ||
          AL.getScopeName()->isStr("intel")))
```
is true rather than using a conditional for it. Then we can remove the code path by which the lambda returns `nullptr` and Klokwork should be pleased again.

Reply:
The removal of the nullptr and adding the assert (such as in [this](https://github.com/intel/llvm/pull/4367/commits/e9f8624b342238dd9fda95e63bd70b661c545b92) commit) has a side effect or two- 
1. it trips up the error checking for the 3 argument requirement for `__attribute__((reqd_work_group_size(3)))` and `[[cl::reqd_work_group_size(3)]]` ; once the suggested assert is in `SetDefault `lambda, there is no opportunity for this check to happen at line 3174 `if (!AL.checkExactlyNumArgs(S, 3))`
2. it asserts everytime it sees a non sycl:: or intel:: usage

Reply:
Hmmm.

I don't see how (1) is correct. Calling `SetDefaultValue()` has no impact on `AL`, so the same checking should happen regardless.

(2) Why would it ever see those usages in the first place? I would expect `isArgExpr()` to return true and you'd never get into `SetDefaultValue()` in that case. Put another way -- if we could have gotten there, then your new asserts will also fail.

(3) While this appeases Klokwork, it doesn't seem like the most clear approach to me. I think making it so that `SetDefaultValue()` cannot return a null pointer is a better approach -- then there's no way to accidentally get a null pointer when assertions are disabled.

I feel like I must be missing something. :-)

Reply:
The OpenCL and cl:: spellings fall outside the scope of sycl:: or intel:: that the assert in SetDefaultValue, hence failing the assert. 

`(AL.getKind() == ParsedAttr::AT_ReqdWorkGroupSize && AL.hasScope() && (AL.getScopeName()->isStr("sycl") || AL.getScopeName()->isStr("intel")))`

In the latest [commit](https://github.com/intel/llvm/pull/4367/commits/7f49a73eb6bbac3d0cdeb50a7462a9ef04592ba0), I placed the error checking for requiring three arguments for OpenCL and cl:: spellings ahead of this assert i.e., before SetDefaultValue lambda. That solves all the aforementioned problems. Please let me know what you think.

Reply:
Your new changes look great!